### ISSUE TITLE
Daylight savings time causes error with time retrieval (off by 1 hour)

### ISSUE THREAD

**Issue by mrchess** [oracle/node-oracledb#414]
Recently I started running into a problem with my `DATE` field.

In the system, my `DATE` is set to `06-NOV-15`.

However, when I query the field the return value is:

`Thu Nov 05 2015 23:00:00 GMT-0500 (EST)`

Notice how it is 1 hour off.

I also noticed that my other date fields seem to return EDT and EST randomly. For example, two other date fields I have returned data like this:

```
...
Thu May 01 2014 00:00:00 GMT-0400 (EDT),
Sun Nov 30 2014 23:00:00 GMT-0500 (EST),
...
```
Notice how one returned EDT and the other EST despite they were both set the same way.

The dates in general have started acting funny ever since DST. Anyone have an idea what may cause this or how to debug it?

**Comment by mrchess** [comment]
I've isolated this into what I believe a bug. Renaming the issue as such. 

Steps to reproduce:
1. Make sure you are in EDT time zone (set your computer to EDT and such).
2. Create a table with a column as a `DATE` field
3. Add two rows any way you wish, eg:

```
INSERT INTO "DATE_BUG" (COLUMN1) VALUES (TO_DATE('2016-04-21 00:00:00', 'YYYY-MM-DD HH24:MI:SS'))

INSERT INTO "DATE_BUG" (COLUMN1) VALUES (TO_DATE('2015-11-06 00:00:00', 'YYYY-MM-DD HH24:MI:SS'))
```
4. Retrieve these rows with the driver

```
var oracledb = require('oracledb');
oracledb.getConnection(
  {
    user          : "system",
    password      : "oracle",
    connectString : "127.0.0.1:1521/orcl"
  },
  function(err, connection)
  {
    connection.execute("SELECT * FROM DATE_BUG",
      {}, 
      function(err, result)
      {
        if (err) { console.error(err.message); return; }
        console.log(result.rows);
      });
  }
);
```

## Expected

It will return two rows, with the dates of:
[ [ Thu Apr 21 2016 00:00:00 GMT-0400 (EDT) ],
  [ Thu Nov 06 2015 00:00:00 GMT-0500 (EST) ] ]

## Actual

The November date is off by 1 hour and it returns:
[ [ Thu Apr 21 2016 00:00:00 GMT-0400 (EDT) ],
  [ Thu Nov 05 2015 23:00:00 GMT-0500 (EST) ] ]

**Comment by cjbj** [comment]
They are being fetched as TIMESTAMP WITH LOCAL TIMEZONE by default.  Note the timezone differs.
If you add `oracledb.fetchAsString = [ oracledb.DATE ];` and set these environment values:
```
export NLS_LANG=.AL32UTF8
export NLS_DATE_FORMAT='YYYY-MM-DD HH24:MI:SS'
```
your script gives: `[ [ '2016-04-21 00:00:00' ], [ '2015-11-06 00:00:00' ] ]`

**Comment by mrchess** [comment]
Interesting. Thanks.

Is environment variables the only way to set those? No way to do it in node?

**Comment by dmcghan** [comment]
Now I'm confused... :)

@mrchess Is the data type of the column DATE or TIMESTAMP WITH LOCAL TIMEZONE? Also, what is the time zone of your database and what is the time zone of your Node.js server?

You can set NLS settings on a connection by executing PL/SQL statements like the following:
"BEGIN EXECUTE IMMEDIATE q'[alter session set NLS_DATE_FORMAT='DD-MM-YYYY']'; END;"

**Comment by dmcghan** [comment]
Basically, the driver converts the timezone of the "date" to your session timezone which defaults to the timezone of the server where you're running Node.js. In my opinion, DATE is an "incomplete" data type as it doesn't contain any timezone related information. I highly recommend the use of TIMESTAMP WITH LOCAL TIME ZONE instead.

I also think that all servers, both database and middle tier (Node.js) should be running in UTC/GMT/Zulu.

If the database only uses TIMESTAMP WITH LOCAL TIME ZONE, then the timezone of the database becomes irrelevant. Then, when you're Node.js server is using UTC, all dates will be normalized to that timezone. This is often what you want as many clients (like browsers) will accept that and then convert to the user's local timezone. In fact, if you do a JSON.stringify on your dates you'll see that they are converted to Zulu (note Z on the end). 

If you are using DATE, then the driver assumes the date was entered with the timezone of your current session - that may or may not be correct. For this reason, DATE values require extra attention. 

I really need to do a write up on all of this... Just need to find the time! :)

**Comment by cjbj** [comment]
Some thoughts
- Date handling hasn't come up as an issue since the fetchAsString / fetchInfo features were introduced
- But https://github.com/oracle/node-oracledb/issues/154 is still open
- I get the impression users are not playing with LOCAL TIME ZONE but just assume the DB and Node are using UTC; and the eventual conversion for the user is done in the browser.  I could be wrong - and not everyone is writing a web app. I'd love more data points.
- Using ALTER SESSION is not good for scalability, since it requires a round-trip to the DB.  Another alternative is to change the session in a trigger but this still requires SQL execution.
- If you set NLS_DATE_FORMAT you also need to set NLS_LANG

**Comment by mrchess** [comment]
@dmcghan Thanks for all the notes on this. To answer your question it was a DATE field. It does indeed looks like a TIMESTAMP insert though. fwiw I used SQLDeveloper to insert a row to a DATE column, and it generated that TIMESTAMP style insert statement as a result.

**Comment by cjbj** [comment]
@mrchess I'll close this now. If you have comments on date handling and how to make things easier while still fitting in with JS, can you add them to #154?

**Comment by bindushastri** [comment]
Hi, In the database(EDT) the date is inserted as 2018-03-07 00:00:00.0
     In the general case of selection am able to fetch correctly.but when I retrieve the data using TRUNC(myDateColumn) the date is pushed back to the previous date ie., 2018-03-06.I appreciate your reply in advance.

**Comment by dmcghan** [comment]
@bindushastri Please open a new issue for your question.

**Comment by bindushastri** [comment]
Sure @dmcghan

**Comment by bindushastri** [comment]
Thank you @dmcghan.
Opened a new issue: Daylight savings time causes an error with time retrieval when TRUNC function is used. #878

**Comment by jaredbaszler** [comment]
Leaving a comment here for anyone else that had to deal with the DST change over the weekend and dates being off by an hour as a result of the info above.  I ended up find a lot of explanation and help from @dmcghan's article o this subject: https://dzone.com/articles/working-with-dates-using-the-nodejs-driver.

The solution that worked because for me was creating a trigger to set the timezone on database logon.  We tied it down to a specific user that was created for my Node apps:

```
CREATE OR REPLACE TRIGGER "node_timezone_set"
AFTER LOGON ON DATABASE
WHEN (LOWER(USER) = 'add username here')
BEGIN
    execute immediate 'alter session set time_zone=''America/Chicago''';
END;
```

This also worked but I didn't want to have to remember to set the time zone every time I created a connection. So we went with the solution above.

```
connection = await oracledb.getConnection(dbConfig);
connection.execute('ALTER SESSION SET TIME_ZONE=\'America/Chicago\'', {}, { autoCommit: true });
```
 Hopefully this will save some others some time.
